### PR TITLE
Issue delete and insert edits when changing parameters

### PR DESCRIPTION
**Pull request by davidwengier**
Part of https://github.com/dotnet/roslyn/issues/59264

I was going to do return type changes with this too, but need to either change SymbolKey, or changes to the SemanticEditInfoComparer: https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs#L3416

### CODE REVIEW

**Comment by davidwengier** [comment]
@tmat PTAL

File: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs [hunk 1]
```diff
@@ -7617,8 +7617,196 @@ static void Main(string[] args)
...
+                {
+                    SemanticEdit(SemanticEditKind.Delete, c => c.GetMembers("C.M").FirstOrDefault(m => m.GetParameterCount() == 1)?.ISymbol, deletedSymbolContainerProvider: c => c.GetMember("C")),
+                    SemanticEdit(SemanticEditKind.Insert, c => c.GetMembers("C.M").FirstOrDefault(m => m.GetParameterCount() == 3)?.ISymbol)
+                },
+                capabilities: EditAndContinueCapabilities.AddMethodToExistingType);
+
+            edits.VerifySemanticDiagnostics(
+                new[]
+                {
+                    Diagnostic(RudeEditKind.InsertNotSupportedByRuntime, "int b", FeaturesResources.parameter),
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:7663]: >InsertNotSupportedByRuntime [](http://example.com/codeflow?start=44&length=27)

As a follow-up, It'd be better to report more specific rude edit here saying that changing signature is not supported by the runtime. Users might not be aware of the fact that we add a new method when they change the signature. It's an implementation detail that shouldn't leak through.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 2]
```diff
@@ -3218,12 +3247,78 @@ public ConstructorEdit(INamedTypeSymbol oldType)
             }
         }
 
+        private static bool TryAddParameterInsertOrDeleteEdits(ArrayBuilder<SemanticEditInfo> semanticEdits, ISymbol containingSymbol, SemanticModel? otherModel, EditAndContinueCapabilitiesGrantor capabilities, Func<SyntaxNode, SyntaxNode?>? syntaxMap, EditScript<SyntaxNode> editScript, CancellationToken cancellationToken, out bool notSupportedByRuntime)
+        {
+            notSupportedByRuntime = false;
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3252]: >notSupportedByRuntime [](http://example.com/codeflow?start=12&length=21)

Consider adding assert on what kinds of containingSymbols are allowed here.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 3]
```diff
@@ -3218,12 +3247,78 @@ public ConstructorEdit(INamedTypeSymbol oldType)
             }
         }
 
+        private static bool TryAddParameterInsertOrDeleteEdits(ArrayBuilder<SemanticEditInfo> semanticEdits, ISymbol containingSymbol, SemanticModel? otherModel, EditAndContinueCapabilitiesGrantor capabilities, Func<SyntaxNode, SyntaxNode?>? syntaxMap, EditScript<SyntaxNode> editScript, CancellationToken cancellationToken, out bool notSupportedByRuntime)
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3250]: >containingSymbol, SemanticModel? otherModel [](http://example.com/codeflow?start=117&length=43)

I'd add doc comments on containingSymbol and otherModel parameters documenting that they are from old and new compilations, or vice versa.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 4]
```diff
@@ -3218,12 +3247,78 @@ public ConstructorEdit(INamedTypeSymbol oldType)
...
+        {
+            notSupportedByRuntime = false;
+
+            // Since we're inserting (or deleting) a parameter node, oldSymbol (or newSymbol) would have been null,
+            // and a symbolkey won't map to the other compilation because the parameters are different, so we have to go back to the edit map
+            // to find the declaration that contains the parameter, and its partner
+
+            // Some sanity checks
+            if (otherModel is null ||
+                containingSymbol.DeclaringSyntaxReferences.Length != 1)
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3260]: >containingSymbol.DeclaringSyntaxReferences.Length != 1 [](http://example.com/codeflow?start=16&length=54)

Does this exclude partial methods? Doc please.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 5]
```diff
@@ -3218,12 +3247,78 @@ public ConstructorEdit(INamedTypeSymbol oldType)
...
 
+        private static bool TryAddParameterInsertOrDeleteEdits(ArrayBuilder<SemanticEditInfo> semanticEdits, ISymbol containingSymbol, SemanticModel? otherModel, EditAndContinueCapabilitiesGrantor capabilities, Func<SyntaxNode, SyntaxNode?>? syntaxMap, EditScript<SyntaxNode> editScript, CancellationToken cancellationToken, out bool notSupportedByRuntime)
+        {
+            notSupportedByRuntime = false;
+
+            // Since we're inserting (or deleting) a parameter node, oldSymbol (or newSymbol) would have been null,
+            // and a symbolkey won't map to the other compilation because the parameters are different, so we have to go back to the edit map
+            // to find the declaration that contains the parameter, and its partner
+
+            // Some sanity checks
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3258]: >Some sanity checks [](http://example.com/codeflow?start=15&length=18)

Not very informative comment.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 6]
```diff
@@ -3218,12 +3247,78 @@ public ConstructorEdit(INamedTypeSymbol oldType)
...
+            {
+                return false;
+            }
+
+            // Now we can work out which is the old and which is the new, depending on which map we found
+            // the match in
+            var oldSymbol = (otherContainingNode == oldNode) ? otherContainingSymbol : containingSymbol;
+            var newSymbol = (otherContainingNode == oldNode) ? containingSymbol : otherContainingSymbol;
+
+            if (CanAddNewMember(oldSymbol, capabilities))
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3298]: >CanAddNewMember [](http://example.com/codeflow?start=16&length=15)

Nit: consider inverting this condition to follow the pattern of if statements above.

**Comment by tmat** [comment]
Reviewed 5 commits. Looks good, just a few minor things.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 7]
```diff
@@ -3218,12 +3247,78 @@ public ConstructorEdit(INamedTypeSymbol oldType)
...
+            }
+
+            var otherContainingSymbol = otherModel.GetDeclaredSymbol(otherContainingNode, cancellationToken);
+            if (otherContainingSymbol is null || !AllowsDeletion(otherContainingSymbol))
+            {
+                return false;
+            }
+
+            // Now we can work out which is the old and which is the new, depending on which map we found
+            // the match in
```
[the remark this answers is not in the document]
Comment (davidwengier) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3294]: The called knows which map the lookup will work in, yes.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 4 above)
Comment (davidwengier) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3260]: This part wasn't a problem with partial methods, so maybe I missed something, but I added a test, and another check to make the test pass :)

**Comment by davidwengier** [comment]
ping @tmat

**Review by tmat — APPROVED** [review]
:shipit:

**Pull request merged**